amethyst_rendy/src/bundle.rs

```diff
@@ -0,0 +1,950 @@
+
+        let mut subpass = SubpassBuilder::new();
+        let mut pass = RenderPassNodeBuilder::new();
+
+        actions.sort_by_key(|a| a.0);
+        for action in actions.drain(..).map(|a| a.1) {
+            match action {
+                RenderableAction::RenderGroup(group) => {
+                    subpass.add_dyn_group(group);
+                }
+            }
+        }
+
+        for (i, color) in outputs.colors.drain(..).enumerate() {
+            match color {
+                OutputColor::Surface(surface, clear) => {
+                    subpass.add_color_surface();
+                    pass.add_surface(surface, clear);
+                }
+                OutputColor::Image(opts) => {
+                    let node = ctx.create_image(opts);
+                    ctx.register_output(TargetImage::Color(self.key, i), node)?;
+                    subpass.add_color(node);
+                }
+            }
+        }
+
+        if let Some(opts) = outputs.depth {
+            let node = ctx.create_image(opts);
+            ctx.register_output(TargetImage::Depth(self.key), node)?;
+            subpass.set_depth_stencil(node);
+        }
+
+        for node in deps {
+            subpass.add_dependency(node);
+        }
+
+        pass.add_subpass(subpass);
+        ctx.submit_pass(self.key, pass)?;
+        Ok(())
+    }
+}
+
+/// An action that represents a single transformation to the
+/// render graph, e.g. addition of single render group.
+///
+/// TODO: more actions needed for e.g. splitting pass into subpasses.
+#[derive(Debug)]
+pub enum RenderableAction<B: Backend> {
+    /// Register single render group for evaluation during target rendering
+    RenderGroup(Box<dyn RenderGroupBuilder<B, Resources>>),
+}
+
+impl<B: Backend> RenderableAction<B> {
+    fn colors(&self) -> usize {
+        match self {
+            RenderableAction::RenderGroup(g) => g.colors(),
+        }
+    }
+
+    fn depth(&self) -> bool {
+        match self {
+            RenderableAction::RenderGroup(g) => g.depth(),
+        }
+    }
+}
+
+/// Trait for easy conversion of various types into `RenderableAction` shell.
+pub trait IntoAction<B: Backend> {
+    /// Convert to `RenderableAction`.
+    fn into(self) -> RenderableAction<B>;
+}
+
+impl<B: Backend, G: RenderGroupBuilder<B, Resources> + 'static> IntoAction<B> for G {
+    fn into(self) -> RenderableAction<B> {
+        RenderableAction::RenderGroup(Box::new(self))
+    }
+}
+
+/// Collection of predefined constants for action ordering in the builtin targets.
+/// Two actions with the same order will be applied in their insertion order.
+/// The list is provided mostly as a comparison point. If you can't find the exact
+/// ordering you need, provide custom `i32` that fits into the right place.
+///
+/// Modules that provide custom render plugins using their own orders should export
+/// similar enum with ordering they have added.
+#[derive(Debug)]
+#[repr(i32)]
+pub enum RenderOrder {
```

Comment:
You leave 9 between first variants to insert custom ones? What will happen if 10 is not enough?

Reply:
It has to be enough. If the order doesn't matter, you put things under the same enum. I can't really think about a case where you need 9 more distinct values inbetween. In fact, I think you would rarely use anything else than defined on that list already.